### PR TITLE
fix(proxy): only save most recent logs from proxy

### LINKED ISSUE #1359: Proxy crashes in long-running app
When leaving the app running (couple of hours) the proxy will crash, likely from the captured logs filling up

```
Proxy process exited with status code ERR_CHILD_PROCESS_STDIO_MAXBUFFER and signal null
```

### PR DESCRIPTION
**Pull request by geigerzaehler**
We change how the proxy process is spawned. Instead of saving all output of the process in a buffer we only save the 200 most recent lines. In addition we forward all of the proxy’s output to the main process output. This allows us to see the proxy logs when starting the app from the terminal.

To implement this we introduce the `ProxyProcessManager` which encapsulates this behavior. The logic for line buffering the output of the process suffers from duplication. I’ve experimented with some abstractions but they were all to complicated and did not add enough value.

Fixes #1359

### CODE REVIEW

File: native/main.ts [hunk 1]
```diff
@@ -145,50 +145,25 @@ const openExternalLink = (url: string): void => {
...
+app.on("ready", () => {
+  proxyProcessManager.run().then(({ status, signal, output: stdout }) => {
     windowManager.sendMessage({
       kind: MainMessageKind.PROXY_ERROR,
       data: {
         status,
         signal,
         stdout,
-        stderr,
+        stderr: "",
```
Comment (rudolfs) [inline native/main.ts:163]: Dev leftover?
I don't see the proxy error logs anymore when I kill the proxy.

Screenshot 2020-11-26 at 13 18 26

File: native/proxy-process-manager.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  // Throws an error if a process is already running.
+  async run(): Promise<ProcessResult> {
+    if (!this.enabled) {
+      return new Promise(() => undefined);
+    }
+
+    if (this.childProcess !== undefined) {
+      throw new Error("Proxy process already started");
+    }
+    const outputBuffer = new CircularBuffer<string>(Array, 200);
```
Comment (rudolfs) [inline native/proxy-process-manager.ts:47]: Is there a reason why we keep it so low? Would opt for `10000`.

**Review by rudolfs — CHANGES_REQUESTED** [review]

File: native/proxy-process-manager.ts (hunk 2 above)
Comment (xla) [reply native/proxy-process-manager.ts:47]: `10000` might be a bit much, these logs are meant to be posted in issues and so on. Maybe somewhere around 500-1000 is managable.

**Comment by geigerzaehler** [comment]
I’ve addressed the issue, @rudolfs, and set the line limit to 1000.

**Comment by rudolfs** [comment]
> I’ve addressed the issue, @rudolfs, and set the line limit to 1000.

https://github.com/radicle-dev/radicle-upstream/pull/1377/files#diff-e9aad0d8c74cb96ae9573c3e3594346000c337b0eab3e680fc0da53b2596e5c8R98

500 !== 1000

File: native/proxy-process-manager.test.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import * as lodash from "lodash";
+import { ProxyProcessManager } from "./proxy-process-manager";
+
+["stdout", "stderr"].forEach(output => {
+  it(`collects and writes ${output}`, async () => {
+    const redirection = output == "stderr" ? "1>&2" : "";
```
Comment (rudolfs) [inline native/proxy-process-manager.test.ts:6]:
```suggestion
    const redirection = output === "stderr" ? "1>&2" : "";
```

**Review by rudolfs — APPROVED** [review]

**Review by xla — APPROVED** [review]
👝 🕠 🃏 💴

**Pull request merged**